### PR TITLE
Handle slow save requests better

### PR DESCRIPTION
**Pull request by philippotto**
- warn the user when there are unsaved changes older than X seconds (for testing, X is set to 5 seconds; for production, I'd suggest to set it to 2 minutes)
- send a warning to airbrake when a save request takes longer than 1 minute (set to 5 s for testing)
- send a a warning to airbrake when there was a conflict state (status code 409)

image

### URL of deployed dev instance (used for testing):
- https://debugsavetimeout.webknossos.xyz

### Steps to test:
- modify a tracing and a warning at the save button should appear after 5 seconds. the back-end has a hardcoded sleep which should make the save request extra slow

### Issues:
- solves front-end part of #4594

------
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes
- [ ] Needs datastore update after deployment
- [x] Ready for review

### CODE REVIEW

File: .circleci/not-on-master.sh [hunk 1]
```diff
@@ -1,8 +1,8 @@
 #!/usr/bin/env bash
 set -Eeuo pipefail
 
-if [ "${CIRCLE_BRANCH}" == "master" ]; then
+#if [ "${CIRCLE_BRANCH}" == "master" ]; then
```
Comment (daniel-wer) [inline .circleci/not-on-master.sh:4]: Reminder to revert this :)

File: frontend/javascripts/oxalis/model/sagas/save_saga.js [hunk 2]
```diff
@@ -49,9 +49,12 @@ import compactSaveQueue from "oxalis/model/helpers/compaction/compact_save_queue
 import compactUpdateActions from "oxalis/model/helpers/compaction/compact_update_actions";
 import messages from "messages";
 import window, { alert, document, location } from "libs/window";
+import ErrorHandling from "libs/error_handling";
 
 import { enforceSkeletonTracing } from "../accessors/skeletontracing_accessor";
 
+const SLOW_SAVE_REQUEST_THRESHOLD = 5 * 1000; // 60 * 1000;
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/model/sagas/save_saga.js:56]: The save interval is 30s, right? Shouldn't we log already if a single request takes longer than 30s, then? Otherwise the save "debt" will continue to build up which could lead to problems.

File: frontend/javascripts/oxalis/model/sagas/save_saga.js [hunk 3]
```diff
@@ -176,16 +179,29 @@ export function* sendRequestToServer(tracingType: "skeleton" | "volume"): Saga<v
...
   while (true) {
     try {
+      const startTime = Date.now();
       yield* call(
         sendRequestWithToken,
         `${tracingStoreUrl}/tracings/${type}/${tracingId}/update?token=`,
         {
           method: "POST",
-          headers: { "X-Date": `${Date.now()}` },
+          // headers: { "X-Date": `${Date.now()}` },
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/model/sagas/save_saga.js:188]: Looks like this can be removed

File: frontend/javascripts/oxalis/model/sagas/save_saga.js [hunk 4]
```diff
@@ -198,6 +214,10 @@ export function* sendRequestToServer(tracingType: "skeleton" | "volume"): Saga<v
         window.onbeforeunload = null;
         yield* call(alert, messages["save.failed_simultaneous_tracing"]);
         location.reload();
+        yield* call(
+          [ErrorHandling, ErrorHandling.notify],
+          new Error("Saving failed due to '409' status code"),
+        );
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/model/sagas/save_saga.js:220]: This should probably be called before the `location.reload()`, shouldn't it. I'd say even before the alert so the request won't get canceled by the reload (does that happen? :thinking:)

File: frontend/javascripts/oxalis/view/action-bar/save_button.js [hunk 5]
```diff
@@ -15,19 +18,33 @@ type OwnProps = {|
...
 |};
 type Props = {| ...OwnProps, ...StateProps |};
 
 type State = {
   isStateSaved: boolean,
+  currentTimestamp: number,
 };
 
 const SAVE_POLLING_INTERVAL = 1000;
+const UNSAVED_WARNING_THRESHOLD = 5 * 1000;
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/action-bar/save_button.js:31]: Reminder to increase this value to two minutes :)

File: frontend/javascripts/oxalis/view/action-bar/save_button.js [hunk 6]
```diff
@@ -15,19 +18,33 @@ type OwnProps = {|
...
+
+const reportUnsavedDurationThresholdExceeded = _.once(() => {
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/action-bar/save_button.js:33]: I'm thinking whether we want to send these periodically (throttled to a larger value like 10 minutes)? But we can also wait and look how often these happen and whether a more detailed logging would help :)

File: webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/controllers/TracingController.scala [hunk 7]
```diff
@@ -115,21 +115,24 @@ trait TracingController[T <: GeneratedMessage with Message[T], Ts <: GeneratedMe
...
-            if (updateGroups.forall(_.transactionGroupCount.getOrElse(1) == 1)) {
-              commitUpdates(tracingId, updateGroups, userToken).map(_ => Ok)
-            } else {
-              updateGroups
-                .foldLeft(tracingService.currentVersion(tracingId)) { (currentCommittedVersionFox, updateGroup) =>
-                  handleUpdateGroupForTransaction(tracingId, currentCommittedVersionFox, updateGroup, userToken)
-                }
-                .map(_ => Ok)
+        for {
+          _ <- Fox.successful(Thread.sleep(1000 * 10))
```
Comment (daniel-wer) [inline webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/controllers/TracingController.scala:119]: Reminder to revert this change :)

File: frontend/javascripts/oxalis/model/sagas/save_saga.js (hunk 2 above)
Comment (philippotto) [reply frontend/javascripts/oxalis/model/sagas/save_saga.js:56]: Good point! I'll change the code to use the existing `PUSH_THROTTLE_TIME` (which is 30s).

File: frontend/javascripts/oxalis/model/sagas/save_saga.js (hunk 4 above)
Comment (daniel-wer) [reply frontend/javascripts/oxalis/model/sagas/save_saga.js:214]: Did you check whether these requests go through? I fear that the reload may abort them, but maybe that doesn't happen. To be on the safe side, you could move it even before the alert call a line above.

Comment (philippotto) [reply frontend/javascripts/oxalis/model/sagas/save_saga.js:213]: You're right, the request is cancelled. I followed your suggestion and moved it which seems to work :)
